diff --git a/configure.ac b/configure.ac
index cbe91cf..447c7e2 100644
--- a/configure.ac
+++ b/configure.ac
@@ -543,7 +543,7 @@ AC_TYPE_SIZE_T
 # Checks for library functions.
 AC_CHECK_FUNCS([gettimeofday memset strchr strcspn strdup strerror strncasecmp strstr strtol strtoul strtoull])
 
-AC_SUBST(FIRMWARE_DIR, "$datadir/sigrok-firmware")
+AC_SUBST(FIRMWARE_DIR, "firmware")
 AC_SUBST(MAKEFLAGS, '--no-print-directory')
 AC_SUBST(AM_LIBTOOLFLAGS, '--silent')
 
