diff --git a/CMakeLists.txt b/CMakeLists.txt
index 443dea6..7b91e0b 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -250,6 +250,7 @@ set(PULSEVIEW_LINK_LIBS
 if(STATIC_PKGDEPS_LIBS)
 	link_directories(${PKGDEPS_STATIC_LIBRARY_DIRS})
 	list(APPEND PULSEVIEW_LINK_LIBS ${PKGDEPS_STATIC_LIBRARIES})
+	list(APPEND PULSEVIEW_LINK_LIBS "-llzma -llcms2") # Quick hack
 else()
 	link_directories(${PKGDEPS_LIBRARY_DIRS})
 	list(APPEND PULSEVIEW_LINK_LIBS ${PKGDEPS_LIBRARIES})
