diff --git a/main.c b/main.c
index 9ab3f77..9d34034 100644
--- a/main.c
+++ b/main.c
@@ -216,7 +216,7 @@ int main(int argc, char **argv)
 		goto done;
 
 	if (opt_pds) {
-		if (srd_init(NULL) != SRD_OK)
+		if (srd_init("decoders") != SRD_OK)
 			goto done;
 		if (srd_session_new(&srd_sess) != SRD_OK) {
 			g_critical("Failed to create new decode session.");
diff --git a/show.c b/show.c
index db643d1..55c1bf6 100644
--- a/show.c
+++ b/show.c
@@ -65,7 +65,7 @@ void show_version(void)
 	printf("\n");
 
 #ifdef HAVE_SRD
-	if (srd_init(NULL) == SRD_OK) {
+	if (srd_init("decoders") == SRD_OK) {
 		printf("Supported protocol decoders:\n");
 		srd_decoder_load_all();
 		for (l = srd_decoder_list(); l; l = l->next) {
