diff -urp libusb-1.0.9.orig/libusb/io.c libusb-1.0.9/libusb/io.c
--- libusb-1.0.9.orig/libusb/io.c	2013-08-28 14:33:15.000000000 +0200
+++ libusb-1.0.9/libusb/io.c	2013-08-28 14:31:25.000000000 +0200
@@ -36,6 +36,14 @@
 
 #include "libusbi.h"
 
+#ifndef TIMESPEC_TO_TIMEVAL
+#define TIMESPEC_TO_TIMEVAL(tv, ts)                                     \
+        do {                                                            \
+                (tv)->tv_sec = (ts)->tv_sec;                            \
+                (tv)->tv_usec = (ts)->tv_nsec / 1000;                   \
+        } while (0)
+#endif
+
 /**
  * \page io Synchronous and asynchronous device I/O
  *
